### PR TITLE
added sampling functionality

### PR DESCRIPTION
**Pull request by thomasfischer11**
includes the tasks (lines from llama2.c): 

8. **Sampler structs, argmax, mult (easy)**
	- lines 577-614
9. Sampler topp (medium)
	- lines 616-665
	- depends on 8
10. Put sampler together (easy)
	- lines 667-714
	- depends on 8, 9

Things that might worry you: 
Used concrete types - expected logits to always be Float32.
I already implemented our own softmax even though someone else had that task, because i needed it. 

🙃

### CODE REVIEW

File: src/sampler.jl [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        if sampler.topp == 0.0 || sampler.topp == 1.0
+            # simply sample from the predicted probability distribution
+            next = sample_mult(probs, coin)
+        else 
+            # top-p (nucleus) sampling, clamping the least likely tokens to zero
+            next = sample_topp(probs, sampler.topp, coin)
+        end
+    end 
+    return next
+end
```
Comment (kleincode) [inline src/sampler.jl:59]: Add docstring

File: src/sampler.jl [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+    return length(probabilities) # in case of rounding errors
+end
+
+function sample_topp(probabilities::Vector{Float32}, topp::Float32, coin::Float32)
+    """
+    Top-p sampling (or "nucleus sampling") samples from the smallest set of
+    tokens that exceed probability topp. This way we never sample tokens that
+    have very low probabilities and are less likely to go "off the rails".
+    Coin is a random number in [0, 1)
+    """
```
Comment (kleincode) [inline src/sampler.jl:90]: Specify range of topp (0 < topp < 1)

File: test/test_sampler.jl [hunk 3]
```diff
@@ -0,0 +1,79 @@
...
+            # 0.7 * 0.4 = 0.28 --> 0.25 < 0.28 < 0.45 --> 7th element
+            sampled_index = sample_topp(probabilities, topp, coin)
+            @test sampled_index == 7  # Should select the 7th element
+        end
+    end
+    @testset "Calling sampler object" begin
+        # Create sampler objects
+        sampler1 = Sampler(0.0, 0.5, 187) # argmax
+        sampler2 = Sampler(0.5, 1.0, 420) # multinomial
+        sampler3 = Sampler(1.0, 0.7, 69) # topp
```
Comment (kleincode) [inline test/test_sampler.jl:48]: nice

File: src/sampler.jl [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+    # truncate the list where cumulative probability exceeds topp
+    cum_prob = 0.0
+    for probind in probindex
+        cum_prob += probind.prob
+        if cum_prob > topp
+            # remove all elements after this one
+            probindex = probindex[1:findfirst(x -> x == probind, probindex)]
+            break # we've exceeded topp 
+        end
+    end
```
[the remark this answers is not in the document]
Comment (kleincode) [reply src/sampler.jl:108]: Edit: If there are multiple occurrences of `probind ` in `probindex`, the current implementation might actually give wrong results.

**Comment by kleincode** [comment]
I think expecting Float32s is fine, as the code for loading the weights from file will also return Float32s.

**Review by kleincode — APPROVED** [review]
top (-p)
